File: frontend/src/lib/masking.ts
/**
 * Frontend-Sicherheitsgürtel for `reveal_participants` (M5c.2, ADR-038).
 *
 * The backend already masks names in `app/services/masking.py` before
 * the REST snapshot leaves the server; this helper re-applies the
 * exact same rule on the client side. It catches two failure modes:
 *
 *  1. A stale TanStack-Query cache that still holds an
 *     un-masked snapshot from before `reveal_participants` was
 *     toggled from `true` to `false`.
 *  2. Future code paths that surface `Person` objects without going
 *     through the REST detail endpoint (e.g. a `Person` RxDB
 *     collection added in a later sub-step). M5c.1b kept Person out
 *     of sync deliberately for that reason — but the helper exists so
 *     the day we add it, the masking rule still holds.
 *
 * The placeholder string mirrors the backend constant and stays in
 * sync via the unit test in `tests/masking.test.ts`.
 */

import type { PersonRead } from "./types";

export const MASK_PLACEHOLDER = "[verborgen]";

interface MaskableEvent {
  reveal_participants: boolean;
}

/**
 * Apply the per-event masking rule to a list of participants.
 *
 * - `reveal_participants === true` → return `participants` as-is.
 * - Otherwise: each participant whose `id === currentPersonId` stays
 *   unmasked; every other participant is rewritten with the
 *   placeholder name and `alias` / `note` cleared.
 *
 * The function is pure; reuse it from any view that renders participant
 * names so the same rule applies whether the data came from REST,
 * RxDB cache, or a future client-side aggregation.
 */
export function maskParticipants(
  participants: readonly PersonRead[],
  event: MaskableEvent,
  currentPersonId: string,
): PersonRead[] {
  if (event.reveal_participants) {
    return [...participants];
  }
  return participants.map((person) => {
    if (person.id === currentPersonId) return person;
    return {
      ...person,
      name: MASK_PLACEHOLDER,
      alias: null,
      note: null,
    };
  });
}

/** Indicator for the UI that an entry is a placeholder, not a real name. */
export function isMasked(person: PersonRead): boolean {
  return person.name === MASK_PLACEHOLDER;
}

File: backend/app/services/masking.py
"""Person-name masking for events with ``reveal_participants=false``.

Per ADR-020 §H, this lives in the service layer because the decision is
per-person and depends on the requester's relationship to the event.
"""

from __future__ import annotations

import uuid

from app.models.event import Event
from app.models.person import Person
from app.schemas.person import PersonRead

PLACEHOLDER = "[verborgen]"


def project_participant(
    person: Person,
    *,
    event: Event,
    requesting_person_id: uuid.UUID,
) -> PersonRead:
    """Return either the real PersonRead or a masked one.

    Rules:
    - ``reveal_participants=true`` -> always real names.
    - The requester sees their own row unmasked.
    - Otherwise the row is masked (name -> placeholder, alias/note dropped).
    """
    base = PersonRead.model_validate(person)
    if event.reveal_participants:
        return base
    if person.id == requesting_person_id:
        return base
    return base.model_copy(update={"name": PLACEHOLDER, "alias": None, "note": None})
